rust/arrow/src/buffer.rs

```diff
@@ -888,6 +888,15 @@ impl MutableBuffer {
         }
         self.len += bytes.len();
     }
+
+    /// Extends the buffer by `len` with all bytes equal to `0u8`, incrementing its capacity if needed.
+    pub fn extend(&mut self, len: usize) {
+        let remaining_capacity = self.capacity - self.len;
+        if len > remaining_capacity {
+            self.reserve(self.len + len);
```

Comment:
should this be `self.reserve(self.len + (len - remaining_capacity));` ?

Reply:
I am sorry, I do not follow: my understanding is that we need to `reserve` so that `self.len < self.capacity` is always true.

If we have a buffer with capacity 64 and `len` 54, and we extend it by 8, we want to ensure that its new capacity is at least `54+8=62`, no?

your formula yields `54 + (8 - (64 - 54)) = 54 + (8 - 10) = 52` (panics with an underflow, I think).

Another way of writing this code would be:
```
self.len += len;
if self.len > self.capacity {
    self.reserve(self.len);
}
```

Reply:
@jorgecarleitao  good spot - what I suggested is more obvious when len > remaining_capacity such if `len = 256` in your example, then requested capacity would be `54 + (256 - (64 - 54)) = 54 + (256 - 10) = 300`; I think it will still work even in your example because the requested of 52 capacity will be lower than existing capacity so no allocation will be performed as it is not necessary because the requested extension of 8 is smaller than remaining capacity of 10.

Reply:
`reserve`s API is to guarantee a capacity of up to the argument, and `extend`'s API is to increase `self.len` by `len`.

If we call `self.reserve(300)`, but then increase the `len` by `256` to `54+256=310`, we will move the pointer past the allocated memory (310 > 300), no?

Reply:
you are right, I misunderstood reserve(x) to mean that we want to increase capacity by x, apologies